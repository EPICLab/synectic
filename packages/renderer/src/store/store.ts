import {combineReducers, configureStore} from '@reduxjs/toolkit';
import branchesReducer from './slices/branches';
import cardsReducer from './slices/cards';
import commitsReducer from './slices/commits';
import filetypesReducer from './slices/filetypes';
import metafilesReducer from './slices/metafiles';
import modalsReducer from './slices/modals';
import reposReducer from './slices/repos';
import stacksReducer from './slices/stacks';
import {listenerMiddleware} from './listenerMiddleware';

export const rootReducer = combineReducers({
  branches: branchesReducer,
  cards: cardsReducer,
  commits: commitsReducer,
  filetypes: filetypesReducer,
  metafiles: metafilesReducer,
  modals: modalsReducer,
  repos: reposReducer,
  stacks: stacksReducer,
});

const store = configureStore({
  reducer: rootReducer,
  devTools: true,
  middleware: getDefaultMiddleware =>
    getDefaultMiddleware().concat([listenerMiddleware.middleware]),
});

export type RootState = ReturnType<typeof rootReducer>;
export type AppDispatch = typeof store.dispatch;

export default {store: store};
