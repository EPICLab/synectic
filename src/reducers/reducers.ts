import { combineReducers } from 'redux';
import {
  // ADD_TODO,
  // TOGGLE_TODO,
  SET_VISIBILITY_FILTER,
  VisibilityFilters,
  ActionType
} from '../actions/actions';

const { SHOW_ALL } = VisibilityFilters;

// type TodoItem = {
//   text: string;
//   completed: boolean;
// }

// type TodoState = {
//   visibilityFilter: string;
//   todos: TodoItem[];
// }

function visibilityFilter(state = SHOW_ALL, action: ActionType) {
  switch (action.type) {
    case SET_VISIBILITY_FILTER:
      return action.filter;
    default:
      return state;
  }
}

// function todos(state: TodoItem[] = [], action: ActionType) {
//   switch (action.type) {
//     case ADD_TODO:
//       return [
//         ...state,
//         {
//           text: action.text,
//           completed: false
//         }
//       ];
//     case TOGGLE_TODO:
//       return state.map((todo, index) => {
//         if (index === action.index) {
//           return Object.assign({}, todo, {
//             completed: !todo.completed
//           });
//         }
//         return todo;
//       });
//     default:
//       return state;
//   }
// }

const todoApp = combineReducers({
  visibilityFilter
});

export default todoApp;