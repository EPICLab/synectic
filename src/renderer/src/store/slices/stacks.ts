import { createEntityAdapter, createSlice, EntityState, PayloadAction } from '@reduxjs/toolkit';
import { DateTime } from 'luxon';
import { Stack } from 'types/stack';

export const stackAdapter = createEntityAdapter<Stack>();

export const stacksSlice = createSlice({
  name: 'stacks',
  initialState: stackAdapter.getInitialState(),
  reducers: {
    stackAdded: stackAdapter.addOne,
    stackRemoved: stackAdapter.removeOne,
    stackUpdated: (state: EntityState<Stack>, action: PayloadAction<Stack>) => {
      stackAdapter.upsertOne(state, {
        ...action.payload,
        modified: DateTime.local().valueOf()
      });
    }
  }
});

export const { stackAdded, stackRemoved, stackUpdated } = stacksSlice.actions;

export default stacksSlice.reducer;
