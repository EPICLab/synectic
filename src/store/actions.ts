import { UUID, Repository, Card, Stack } from './types';

export enum ActionKeys {
  INITIALIZE_CANVAS,
  ADD_REPO, REMOVE_REPO, UPDATE_REPO,
  ADD_CARD, REMOVE_CARD, UPDATE_CARD,
  ADD_STACK, REMOVE_STACK, UPDATE_STACK
}

export type Actions =
  | InitializeCanvasAction
  | AddRepoAction
  | RemoveRepoAction
  | UpdateRepoAction
  | AddCardAction
  | RemoveCardAction
  | UpdateCardAction
  | AddStackAction
  | RemoveStackAction
  | UpdateStackAction;

type InitializeCanvasAction = {
  type: ActionKeys.INITIALIZE_CANVAS;
};

type AddRepoAction = {
  type: ActionKeys.ADD_REPO;
  id: UUID;
  repo: Repository;
};

type RemoveRepoAction = {
  type: ActionKeys.REMOVE_REPO;
  id: UUID;
}

type UpdateRepoAction = {
  type: ActionKeys.UPDATE_REPO;
  id: UUID;
  repo: Partial<Repository>;
};

type AddCardAction = {
  type: ActionKeys.ADD_CARD;
  id: UUID;
  card: Card;
};

type RemoveCardAction = {
  type: ActionKeys.REMOVE_CARD;
  id: UUID;
};

type UpdateCardAction = {
  type: ActionKeys.UPDATE_CARD;
  id: UUID;
  card: Partial<Card>;
};

type AddStackAction = {
  type: ActionKeys.ADD_STACK;
  id: UUID;
  stack: Stack;
}

type RemoveStackAction = {
  type: ActionKeys.REMOVE_STACK;
  id: UUID;
};

type UpdateStackAction = {
  type: ActionKeys.UPDATE_STACK;
  id: UUID;
  stack: Partial<Stack>;
};