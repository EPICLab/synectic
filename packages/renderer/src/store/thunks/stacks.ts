import {uuid} from '#preload';
import type {UUID} from '@syn-types/app';
import {DateTime} from 'luxon';
import {createAppAsyncThunk} from '../hooks';
import cardSelectors from '../selectors/cards';
import stackSelectors from '../selectors/stacks';
import {cardUpdated} from '../slices/cards';
import {stackAdded, stackRemoved, stackUpdated} from '../slices/stacks';

/**
 * Create a new Stack component that is added to the Redux store and the UI.
 * @param name The name of the new Stack component.
 * @param cards The UUIDs of Card components that should be pushed onto the Stack component.
 * @param x The horizontal position of the new Stack component; defaults to 40.
 * @param y The vertical position of the new Stack component; defaults to 40.
 * @returns {UUID} The UUID of the new Stack component.
 */
export const createStack = createAppAsyncThunk<
  UUID,
  {name: string; cards: UUID[]; x?: number; y?: number}
>('stacks/createStack', async (input, thunkAPI) => {
  const cards = cardSelectors.selectByIds(thunkAPI.getState(), input.cards);

  const stackId = uuid();
  thunkAPI.dispatch(
    stackAdded({
      id: stackId,
      name: input.name,
      created: DateTime.local().valueOf(),
      modified: DateTime.local().valueOf(),
      cards: input.cards,
      x: input.x ?? 40,
      y: input.y ?? 40,
    }),
  );
  cards.map(card =>
    thunkAPI.dispatch(
      cardUpdated({
        ...card,
        captured: stackId,
      }),
    ),
  );
  return stackId;
});

/**
 * Remove a Stack component from the Redux store and the UI.
 * @param stack The UUID of the Stack component to remove.
 */
export const removeStack = createAppAsyncThunk<void, {stack: UUID}>(
  'stacks/removeStack',
  async (input, thunkAPI) => {
    const state = thunkAPI.getState();
    const stack = stackSelectors.selectById(state, input.stack);

    if (stack) {
      const cards = cardSelectors.selectByIds(state, stack.cards);
      cards.map((card, index) =>
        thunkAPI.dispatch(
          cardUpdated({
            ...card,
            captured: undefined,
            x: stack.x + index * 15,
            y: stack.y + (index + 1) * 15,
          }),
        ),
      );
      thunkAPI.dispatch(stackRemoved(stack.id));
    }
  },
);

/**
 * Push Card components onto the Stack by updating the Redux store and UI.
 * @param stack The UUID of the Stack component to receive the cards.
 * @param cards The UUIDs of Card components that should be pushed onto the Stack component.
 */
export const pushCards = createAppAsyncThunk<void, {stack: UUID; cards: UUID[]}>(
  'stacks/pushCards',
  async (input, thunkAPI) => {
    const state = thunkAPI.getState();
    const stack = stackSelectors.selectById(state, input.stack);
    const cards = cardSelectors.selectByIds(state, input.cards);

    if (stack && cards.length > 0) {
      thunkAPI.dispatch(
        stackUpdated({
          ...stack,
          cards: [...stack.cards, ...input.cards],
        }),
      );
      cards.map(card =>
        thunkAPI.dispatch(
          cardUpdated({
            ...card,
            captured: stack.id,
          }),
        ),
      );
    }
  },
);

/**
 * Pop Card components off of the Stack by updating the Redux store and UI.
 * @param stack The UUID of the Stack component to pop the cards from.
 * @param card The UUID of the Card component to be popped from the Stack component.
 * @param x The updated horizontal position of the popped Card component.
 * @param y The vertical position of the popped Card component.
 */
export const popCard = createAppAsyncThunk<void, {stack: UUID; card: UUID; x: number; y: number}>(
  'stacks/popCard',
  async (input, thunkAPI) => {
    const state = thunkAPI.getState();
    const stack = stackSelectors.selectById(state, input.stack);
    const card = cardSelectors.selectById(state, input.card);

    if (stack && card) {
      if (stack.cards.length <= 2) {
        thunkAPI.dispatch(removeStack({stack: stack.id}));
      } else {
        thunkAPI.dispatch(
          stackUpdated({
            ...stack,
            cards: stack.cards.filter(id => id !== input.card),
          }),
        );
      }
      thunkAPI.dispatch(
        cardUpdated({
          ...card,
          captured: undefined,
          x: input.x,
          y: input.y,
        }),
      );
    }
  },
);
