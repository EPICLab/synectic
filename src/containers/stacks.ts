import { DateTime } from 'luxon';
import { v4 } from 'uuid';

import { Card, Stack } from '../types';
import { ActionKeys, NarrowActionType } from '../store/actions';
import { updateCard } from './cards';
import { removeItemInArray } from '../store/immutables';
import { XYCoord } from 'react-dnd';

type AddStackAction = NarrowActionType<ActionKeys.ADD_STACK>;
type UpdateStackAction = NarrowActionType<ActionKeys.UPDATE_STACK>;
type UpdateCardAction = NarrowActionType<ActionKeys.UPDATE_CARD>;

/**
 * Action Creator for composing a valid ADD_STACK Redux Action, which requires additional UPDATE_CARD Redux Actions for updating all 
 * child cards captured within the new Stack.
 * @param name Name of the new Stack object.
 * @param cards Array of at least two Card objects to be contained with the new Stack.
 * @param note (Optional) Note field related to the new Stack.
 * @return An array of `AddStackAction` and `UpdateCardAction` objects that can be dispatched via Redux.
 */
export const addStack = (name: string, cards: Card[], note?: string): (AddStackAction | UpdateCardAction)[] => {
  const stack: Stack = {
    id: v4(),
    name: name,
    created: DateTime.local(),
    modified: DateTime.local(),
    note: note ? note : '',
    cards: cards.map(card => card.id),
    left: cards[0].left,
    top: cards[0].top
  };
  const actions: (AddStackAction | UpdateCardAction)[] = [{ type: ActionKeys.ADD_STACK, id: stack.id, stack: stack }];
  cards.map((card, index) => actions.push(updateCard({ ...card, captured: stack.id, top: (10 * index + 50), left: (10 * index + 10) })));
  return actions;
}

/**
 * Action Creator for composing a valid UPDATE_STACK Redux Action. If the current Redux store does not contain a matching stack 
 * (based on UUID) for the passed parameter, then dispatching this action will not result in any changes in the Redux store state.
 * @param stack A `Stack` object containing new field values to be updated.
 * @return An `UpdateStackAction` object that can be dispatched via Redux, including an updated timestamp in the `modified` field.
 */
export const updateStack = (stack: Stack): UpdateStackAction => {
  return {
    type: ActionKeys.UPDATE_STACK,
    id: stack.id,
    stack: { ...stack, modified: DateTime.local() }
  };
};

/**
 * Action Creator for composing a valid UPDATE_STACK and UPDATE_CARD Redux actions for capturing a child card and containing
 * that card within a stack. Positioning of the card becomes relative to the bounds of the stack.
 * @param stack The target `Stack` object that should receive the new card.
 * @param card The `Card` object to be added to the stack.
 * @return An array of `UpdateStackAction` and `UpdateCardAction` objects that can be dispatched via Redux.
 */
export const appendCard = (stack: Stack, card: Card): (UpdateStackAction | UpdateCardAction)[] => {
  return [
    updateStack({ ...stack, cards: [...stack.cards, card.id] }),
    updateCard({ ...card, captured: stack.id, top: (10 * stack.cards.length + 50), left: (10 * stack.cards.length + 10) })
  ];
}

/**
 * Action Creator for composing a valid UPDATE_STACK and UPDATE_CARD Redux actions for removing a child card contained within
 * a stack. If the current Redux store does not contain a matching card (based on UUID) contained in the stack passed as a
 * parameter, then dispatching this action will not result in any changes in the Redux store state. Positioning of the card
 * becomes relative to the bounds of the canvas.
 * @param stack The target `Stack` object that contains the card to be removed.
 * @param card The `Card` object to be removed from the stack.
 * @param delta The { x, y } difference between the last recorded client offset of the pointer and the client offset when the 
 * current drag operation has started (gathered from `DropTargetMonitor.getDifferenceFromInitialOffset()`).
 * @return An array of `UpdateStackAction` and `UpdateCardAction` objects that can be dispatched via Redux.
 */
export const removeCard = (stack: Stack, card: Card, delta: XYCoord): (UpdateStackAction | UpdateCardAction)[] => {
  return [
    updateStack({ ...stack, cards: removeItemInArray(stack.cards, card.id) }),
    updateCard({
      ...card,
      captured: undefined,
      left: Math.round(stack.left + card.left + delta.x),
      top: Math.round(stack.top + card.top + delta.y)
    })
  ];
}