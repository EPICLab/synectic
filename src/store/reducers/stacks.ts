import type { Stack } from '../../types';
import { Action, ActionKeys } from '../actions';
import { addItemInMap, removeItemInMap, updateItemInMapById, updateObject } from '../immutables';

export const stackReducer = (state: { [id: string]: Stack } = {}, action: Action): { [id: string]: Stack } => {
  switch (action.type) {
    case ActionKeys.ADD_STACK:
      return addItemInMap(state, action.stack);
    case ActionKeys.REMOVE_STACK:
      return removeItemInMap(state, action.id);
    case ActionKeys.UPDATE_STACK:
      return updateItemInMapById(state, action.id, (stack => updateObject(stack, action.stack)));
    default:
      return state;
  }
}