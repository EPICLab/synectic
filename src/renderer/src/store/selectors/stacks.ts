import { stackAdapter } from '../slices/stacks';
import type { RootState } from '../store';

const selectors = stackAdapter.getSelectors<RootState>(state => state.stacks);

const stackSelectors = { ...selectors };

export default stackSelectors;
