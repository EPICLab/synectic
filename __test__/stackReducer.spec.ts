import { DateTime } from 'luxon';

import { Stack } from '../src/store/types';
import { ActionKeys } from '../src/store/actions';
import { stackReducer } from '../src/store/reducers/stacks';

describe('stackReducer', () => {
  const stacks: { [id: string]: Stack } = {
    '141': {
      id: '141',
      name: 'testStack',
      created: DateTime.fromISO('2019-11-19T19:22:47.572-08:00'),
      modified: DateTime.fromISO('2019-11-19T19:22:47.572-08:00'),
      note: 'This stack has all of my important API documentation.',
      cards: ['32', '14'],
      left: 16,
      top: 21
    }
  }

  const newStack: Stack = {
    id: '2',
    name: 'stack',
    created: DateTime.fromISO('2016-01-25T11:38:11.720-08:00'),
    modified: DateTime.fromISO('2016-01-27T19:05:09.845-08:00'),
    note: 'This stack contains very important cards for Project X',
    cards: [], left: 78, top: 13
  }

  it('stackReducer appends a new stack to state on action ADD_STACK', () => {
    const addedStacks = stackReducer(stacks, { type: ActionKeys.ADD_STACK, id: newStack.id, stack: newStack });
    expect(Object.keys(stacks)).toHaveLength(1);
    expect(Object.keys(addedStacks)).toHaveLength(2);
  });

  it('stackReducer removes a stack from state on action REMOVE_STACK', () => {
    const matchedStacks = stackReducer(stacks, { type: ActionKeys.REMOVE_STACK, id: '141' });
    expect(Object.keys(matchedStacks)).toHaveLength(0);
  });

  it('stackReducer resolves non-matching stack in state on action REMOVE_STACK', () => {
    const nonMatchedStacks = stackReducer(stacks, { type: ActionKeys.REMOVE_STACK, id: '99' });
    expect(Object.keys(nonMatchedStacks)).toHaveLength(Object.keys(stacks).length);
  });

  it('stackReducer updates state of matched stack on action UPDATE_STACK', () => {
    const updatedStacks = stackReducer(stacks, {
      type: ActionKeys.UPDATE_STACK, id: '141', stack: {
        name: 'subtestStack',
        note: 'A subset of our favorite API doc blocks.'
      }
    });
    expect(updatedStacks).not.toMatchObject(stacks);
    expect(updatedStacks).toMatchSnapshot();
  });
});