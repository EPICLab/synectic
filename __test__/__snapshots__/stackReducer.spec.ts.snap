// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`stackReducer stackReducer updates state of matched stack on action UPDATE_STACK 1`] = `
Object {
  "141": Object {
    "cards": Array [
      "32",
      "14",
    ],
    "created": "2019-11-19T19:22:47.572-08:00",
    "id": "141",
    "left": 16,
    "modified": "2019-11-19T19:22:47.572-08:00",
    "name": "subtestStack",
    "note": "A subset of our favorite API doc blocks.",
    "top": 21,
  },
}
`;
